### DISCUSSION TITLE
Question about the EC demo

### DISCUSSION THREAD

**Discussion by chao0x11** [openssl/openssl#26083] [Q&A] [answered]
Hello,

I'm learning the demo of openssl. But I'm blocked in EC keygen and signature. In [openssl-3.4.0/demos/pkey/EVP_PKEY_EC_keygen.c ](https://fossies.org/linux/openssl/demos/pkey/EVP_PKEY_EC_keygen.c), the maximal size of public key and private key are 80 (line 89 and line 90). But in [openssl-3.4.0/demos/signature/EVP_EC_Signature_demo.h](https://fossies.org/linux/openssl/demos/signature/EVP_EC_Signature_demo.h) The size of public key and private key seems much larger than 80. How can I use keygen.c to generate keys and use signature.c to sign file with that generated keys. 

Thanks.

**Comment by t8m** [comment, marked as answer]
The keys in the signature demo are DER-encoded EC keys with explicit parameters.

The values printed in get_key_values() are the individual public key and private key values (in separate buffers). The parameters are also implicit via the named curve.

You would have to actually use the encoder (or the PEM functions) to create an encoding of the private key and a public key to files that could be later read and used for signing or verifying signatures.

**Reply by chao0x11** [reply]
Thanks, it helps me.

**Reply by chao0x11** [reply]
I also have a question. I use DER-encoder to encode my key. The size of the result key is 118 (passphrase != NULL) and 88 (passphrase == NULL). How can I get the same length of key in [EVP_EC_Signature_demo.h](https://fossies.org/linux/openssl/demos/signature/EVP_EC_Signature_demo.h)? 
p.s., I use secp256k1 in my test. Should I change the curve?
